policy/compiler/compiler.go

```diff
@@ -335,6 +335,41 @@ func (tc *templateCompiler) compileOpenAPISchema(dyn *model.DynValue,
 	if found {
 		schema.DefaultValue = tc.convertToSchemaType(elem.Ref.ID, elem.Ref.Value, schema)
 	}
+
+	tc.validateOpenAPISchema(dyn, schema)
+}
+
+func (tc *templateCompiler) validateOpenAPISchema(dyn *model.DynValue, schema *model.OpenAPISchema) {
+	m := tc.mapValue(dyn)
+	if len(m.Fields) > 0 && len(schema.Type) == 0 {
+		tc.reportErrorAtID(dyn.ID,
+			"invalid schema. fields not empty, type must be set.")
+	}
+	if len(schema.Properties) > 0 && schema.AdditionalProperties != nil {
+		tc.reportErrorAtID(dyn.ID,
+			"invalid schema. properties set, additionalProperties must not be set.")
+	}
+	if (len(schema.Properties) > 0 || schema.AdditionalProperties != nil) &&
```

Comment:
In order to provide a higher fidelity signal about where the error occurs, I would test the properties and additional properties separately and report separately:

```go
p, hasProps := m.GetField("properties")
if hasProps && schema.Type != "object" {
   t.reportErrorAtID(p.Ref.ID, ...)
}
ap, hasAdditionalProps := m.GetField("additionalProperties")
if hasAdditionalProps && schema.Type != "object" {
  ...
}
if hasProps && hasAdditionalProps {
   // report error at additional properties.
}
```

You can then reuse the `hasProps` and `hasAdditionalProps` booleans within the `schema.Items` tests to report an error on the `items` field if either flag is true.